packages/wasm/crate/src/planner.rs

```diff
@@ -525,36 +525,69 @@ pub async fn plan_transaction_inner<Db: Database>(
-    //             amount_to_spend: None,
-    //         })
-    //         .await?;
-    //     notes_by_asset_id.insert(
-    //         asset_id.try_into()?,
-    //         prioritize_and_filter_spendable_notes(records.clone()),
-    //     );
-
-    //     for record in records {
-    //         // filter zero-valued notes from SNR set
-    //         if record.clone().note.amount() != 0u64.into() {
-    //             let spend: SpendPlan =
-    //                 SpendPlan::new(&mut OsRng, record.clone().note, record.clone().position);
-    //             actions_list.push(spend);
-    //         }
-    //     }
-    // }
+    for tpr::Spend { value, address } in request.spends.clone() {
+        let value = value.ok_or_else(|| anyhow!("missing value in spend"))?;
+        let address = address.ok_or_else(|| anyhow!("missing address in spend"))?;
+        let asset_id = value
+            .asset_id
+            .ok_or_else(|| anyhow!("missing asset_id in spend"))?;
+
+        // Constraint: validate the transaction planner request is constructed with a single spend request.
+        if request.spends.len() > 1 {
+            let error_message =
+                "Invalid transaction: The transaction was constructed improperly.".to_string();
+            return Err(WasmError::Anyhow(anyhow!(error_message)));
+        }
+
+        // Find all the notes of this asset in the source account.
+        let records = storage
+            .get_notes(NotesRequest {
+                include_spent: false,
+                asset_id: Some(asset_id.clone()),
+                address_index: Some(source_address_index.into()),
+                amount_to_spend: None,
+            })
+            .await?;
+
+        // Accumlate the total available note balance for the asset id.
+        let accumulated_note_amounts = records
+            .iter()
+            .map(|record| record.note.value().amount)
+            .fold(penumbra_num::Amount::zero(), |acc, note_value| {
+                acc + note_value
+            });
+
+        // Constraint: check if the requested spend amount is equal to the accumulated note balance.
+        if accumulated_note_amounts.to_proto() != value.amount.unwrap() {
+            let error_message =
+                "Invalid transaction: The transaction was constructed improperly.".to_string();
+            return Err(WasmError::Anyhow(anyhow!(error_message)));
+        }
+
+        for record in records.clone() {
+            // Filter out zero-valued notes from spendable note record (SNR) set.
+            if record.clone().note.amount() != 0u64.into() {
+                let spend: SpendPlan =
+                    SpendPlan::new(&mut OsRng, record.clone().note, record.clone().position);
+                actions_list.push(spend);
+            }
+        }
+
+        // Constraint: validate that each spend action's asset ID matches the fee asset ID.
+        for action in actions_list.actions() {
```

Comment:
good question; 

Given the sensitivity of this feature, we want to make it as restrictive as possible by peppering many constraint checks. This constraint is specifically _only_ relevant and scoped for a [Spend TPR](https://github.com/penumbra-zone/web/pull/1801/files#diff-ec51d55a0b0ab2ce263357c3d43509ddfe1d88ace8bc963213a0899cca79a0f0R528). It enforces that all the notes being sent to the recipient are of the same denomination and that fees are deducted from those notes. For example, a user can only spend the maximum amount of UM or USDC, but not send both max USDC and the fee note UM to the recipient, which was the source of the previous bug. I'll add a code comment.

Therefore, we don't want to relocate the constraint check by itself below all actions, since again it only applies to spend transaction planner requests.